packages/router/src/navigation_transition.ts

```diff
@@ -602,37 +603,67 @@ export class NavigationTransitions {
-                         }),
-
-                         activateRoutes(
-                             this.rootContexts, router.routeReuseStrategy,
-                             (evt: Event) => this.events.next(evt), this.inputBindingEnabled),
-
                          // Ensure that if some observable used to drive the transition doesn't
                          // complete, the navigation still finalizes This should never happen, but
                          // this is done as a safety measure to avoid surfacing this error (#49567).
                          take(1),
 
-                         tap({
-                           next: (t: NavigationTransition) => {
+                         switchMap(t => {
+                           // Everything in the transition must happen in a synchronous block.
+                           // Breaking it up results in different timing of router state updates and
+                           // router events.
+                           const transitionDone = this.transitionHelper(() => {
+                             // Because startViewTransition happens asynchronously, we need to
+                             // ensure that we are on the same navigation when the callback
+                             // executes.
+                             if (this.currentNavigation?.id !== t.id) {
+                               return;
+                             }
+                             const targetRouterState = createRouterState(
+                                 router.routeReuseStrategy, t.targetSnapshot!,
+                                 t.currentRouterState);
+                             this.currentTransition = t =
+                                 overallTransitionState = {...t, targetRouterState};
+
+                             this.events.next(new BeforeActivateRoutes());
+                             // The above event causes entrance into user code that may error and
+                             // cancel the navigation. Ensure we are still processing before
+                             // continuing.
+                             if (this.currentNavigation?.id !== t.id) {
+                               return;
+                             }
+
+                             // activate and deactivate routes via the router outlets
+                             new ActivateRoutes(
+                                 router.routeReuseStrategy, targetRouterState, t.currentRouterState,
+                                 (evt: Event) => this.events.next(evt), this.inputBindingEnabled)
+                                 .activate(this.rootContexts);
+                             // Activating routes enters user code that could trigger a new
+                             // navigation. Ensure the current navigation is the same as this
+                             // transition before continuing.
+                             if (this.currentNavigation?.id !== t.id) {
```

Comment:
nit: we can also have a helper function here that would act as an extra inline documentation, something like this:
```suggestion
                             if (!this.wasTransitionInterrupted(t.id)) {
```

Reply:
Updated approach to minimize changes in the navigation transition.